livebook_examples/speech_to_text/speech_to_text.livemd

```diff
@@ -0,0 +1,172 @@
+
+### Buffers handling
+
+Once a buffer arrives, we store its payload along all already seen audio samples in the element's state.
+
+The audio needs to be split into chunks and the transcription is performed on each of these chunks.
+With the `@chunk_duration` attribute we can specify how long the chunk should be.
+
+In case there are enough samples already stored in the element's state and no transcription occurs at a time, the transcription is launched as an asynchronous task. Once its results are available, a message of form `{:transcribed, transcription}` is sent to the element.
+
+We reflect these requirements in the definition of the accepted input stream format: `Membrane.RawAudio`.
+
+```elixir
+alias Membrane.RawAudio
+
+defmodule SpeechToText do
+  use Membrane.Sink
+
+  @sample_rate 16_000
+  # sample format is `:f32le`, so each format is written on 32 bits = 4 bytes
+  @bytes_per_sample 4
+  @channels 1
+  # seconds
+  @chunk_duration 5
+  @chunk_size @sample_rate * @bytes_per_sample * @channels * @chunk_duration
+
+  def_input_pad(:input,
+    accepted_format: %RawAudio{
+      channels: @channels,
+      sample_rate: @sample_rate,
+      sample_format: :f32le
+    },
+    demand_mode: :auto
+  )
+
+  @impl true
+  def handle_init(_ctx, _opts) do
+    {:ok, whisper} = Bumblebee.load_model({:hf, "openai/whisper-tiny"})
+    {:ok, featurizer} = Bumblebee.load_featurizer({:hf, "openai/whisper-tiny"})
+    {:ok, tokenizer} = Bumblebee.load_tokenizer({:hf, "openai/whisper-tiny"})
+    {:ok, generation_config} = Bumblebee.load_generation_config({:hf, "openai/whisper-tiny"})
+
+    serving =
+      Bumblebee.Audio.speech_to_text(whisper, featurizer, tokenizer, generation_config,
+        defn_options: [compiler: EXLA]
+      )
+
+    state = %{serving: serving, samples: <<>>, transcribing?: false}
+    {[], state}
+  end
+
+  @impl true
+  def handle_write(:input, buffer, _ctx, state) do
+    all_samples = state.samples <> buffer.payload
+
+    if byte_size(all_samples) > @chunk_size and not state.transcribing? do
+      transcribe(all_samples, state.serving)
+      {[], %{state | samples: <<>>, transcribing?: true}}
+    else
+      {[], %{state | samples: all_samples}}
+    end
+  end
+
+  defp transcribe(data, serving) do
+    send_to = self()
+
+    Task.async(fn ->
+      model_input = Nx.from_binary(data, :f32)
+      results = Nx.Serving.run(serving, model_input)
+      transcription = Enum.map(results.results, & &1.text) |> Enum.join()
+      send(send_to, {:transcribed, transcription})
+    end)
+  end
+
+  @impl true
+  def handle_info({:transcribed, transcription}, _ctx, state) do
+    IO.inspect(transcription, label: "Transcription")
+    state = %{state | transcribing?: false}
+    {[], state}
+  end
+
+  @impl true
+  def handle_info(_other, _ctx, state) do
+    {[], state}
+  end
```

Comment:
so that the element won't crash when it receives some other message than `{:transcribed, transcription}`. 
I just assume it a good practice to add such a general case to handle some unexpected messages.

Reply:
done